### ISSUE TITLE
Codebase Impovement Proposals

### ISSUE THREAD

**Issue by brusherru** [spacemeshos/smapp#799]
Right now, we have many small issues within our codebase, which makes the code less maintainable and robust.
This issue contains a quick overview of problems and proposals related to some topics with a short description.

Feel free to participate in discussions and make your proposals.
In the end, we should create small issues based on this one.

Topics:

1. [Redux as a single source of truth](#p1)
2. [Atomic state changes](#p2)
3. [Add data types everywhere](#p3)
4. [Tweak eslint rules](#p4)
5. [Repo structure](#p5)
6. [Sync App settings with redux state](#p6)
7. [Tests](#p7)

In the discussion please refer to the problem/proposal number.

---

<a name="p1"></a>

## 1. Redux as a single source of truth
We have a mix of redux state with component states, that relates to the whole App. It causes some problems, that can not be fixed easily, or, at least, the logic is not straightforward.
For example, we use `react-router`, which contains the route in its component state and passes history and location in Context. So it works in parallel with redux, and it can not be handled from redux actions/middlewares. As a result, we have to add some return value to actions, check them, and then decide which screen to show within components. Also, it makes error handling a bit complicated.

### Proposal №1
Use component state only for things that are related only to this component and which are not actual after component unmounted. For example, it is good to store an input value until it will commit to the redux state, or sent it to some handler, which, in the end, will update the redux state.
For other cases use redux state.

This proposal includes replacing `react-router` with [react-redux-router](https://github.com/reactjs/react-router-redux) or something similar. So we can trigger changing of the screen by changing the state right within redux.

### Proposal №2
Actions should be simple and don't know anything about consumers. So they should not return anything except simple redux `Action` (object), which should be dispatched.
In case that some thunk action depends on the result of another `ThunkAction` it should get action, which can be dispatched, and take needed data from the payload.

For example:
User creates new transaction and dispatches a thunk Action `sendTransactionAndOpenDetails`, which calls
- `sendTransaction :: (args) -> { type: 'SEND_TRANSACTION', payload: { txId: TransactionID } }`
- and then it dispatches returned action, takes a transaction id from the payload, and dispatches switching screen (`react-redux-router`) to `/transaction/details/${txId}`

### Proposal №3
Since this topic needs a huge refactoring we can also switch to some handy libraries from `redux-toolkit`, like `redux-slice` to make less boilerplate and make the related code more maintainable.

---

<a name="p2"></a>

## 2. Atomic state changes
Most of the complicated actions dispatch a bunch of small actions that update redux state one by one.
Redux has an internal mechanism that reduces the amount of re-renderings for such cases, so probably this is not a problem. 
However, if something will go wrong on one of the dispatched actions we'll get an inconsistent state: something already updated, something not.

### Proposal
To keep the App state consistent we should make state changes atomic. It means, that if some action will change a bunch of things in the state it should be done within one change. So if something went wrong the state won't change at all and it will be kept in a previous consistent state.

Redux community already has solutions for such cases, which simplifies writing small actions and batching it: [redux-batched-actions](https://github.com/tshelburne/redux-batched-actions).
But I'm not sure that this solution will fit all our needs:
> You can add a middleware to dispatch each of the bundled actions. This can be used if other middlewares are listening for one of the bundled actions to be dispatched.

More research is needed.

---

<a name="p3"></a>

## 3. Add data types everywhere
In the code, we can often meet `Any` type or `@ts-ignore`. It makes code more complicated and turns off static typing. As a result, it lowers the power of TypeScript.
Such things also hide errors that are easily caught by static typing.

### Proposal №1
Get rid of all `Any` types and `@ts-ignore`.
As a result, we can turn on `strict` mode on TS and `@typescript-eslint/ban-ts-ignore` on eslint. It will make code more robust and maintainable.

Also, it includes passing types through IPC and other pub/sub buses. And one more thing, which I highlighted in a separate proposal:

### Proposal №2
**✅ Done**: https://github.com/spacemeshos/smapp/pull/809
Generate typescript types from `.proto` files and refactor Services a bit. I've already taken a look at few libraries which can help up with that. And I think, that just updating `@grpc/proto-loader` to the latest one and use bundled `proto-loader-gen-types` will work for us and require the smallest amount of changes to the code base.

---

<a name="p4"></a>

## 4. Tweak eslint rules
Probably this is the most minor thing (except ban `@ts-ignore` described above), but it may make code more readable and also provide better git diffs.

### Proposal
Mark most things as errors to avoid collecting linting warnings.
And change rules:
- disallow `@ts-ignore`
- limit line length to 80-120 chars
- allow `no-parens` for small arrow functions. E.G. `then(a => a.payload)`
- turn on `comma-dangle` (clean diffs on property set change)

I believe that this is not a full list of changes :)

---

<a name="p5"></a>

## 5. Repo structure
We have a little mess in the directory structure:
- `/desktop` — main process files
- `/desktop/dist` — built renderer files
- `/app` — renerer process sources
- `/resources` — assets for electron-builder
- `/shared` — sources of things shared between the renderer and main processes
- `/dll` — built js files for dev
- `/release` — bundled electron App

So all these things look complicated since we also have `scripts`, `configs`, `proto` directories. Moreover, we have a bunch of webpack configs which point to different paths. And complicated `.gitignore` file.

### Proposal
- Move all sources and assets in the `src` directory
  - `/src/main`
  - `/src/renderer`
  - `/src/assets`
  - `/src/shared`
- Put all artifacts in the `dist` folder:
  - `/dekstop/dist` -> `/dist/build`
  - `/dll` -> `/dist/dev`
  - `release` -> `/dist/release`
- Update configs (and get rid of excessive ones), `.gitignore`

---

<a name="p6"></a>

## 6. App Settings
Right now we store in the App settings only few things, such as `netSettings` record, `localNode` flag, `accounts` map. It should contain more things.

E.G. Resizing the App window, or switching theme, does not save between App sessions.

### Proposal
Create a separate redux reducer, which will be synced with these settings. On the startup, it can request settings from the main process and update the state. On any further updating of this state (it can be caught by middleware), it can request updating such data in the electrons persistent store.

---

<a name="p7"></a>

## 7. Tests
Now it has no tests at all.
In general, this can be understood, because basically, our application is a facade for interacting with the Node. So most of the things are depend on Node, and to make most of the tests we have to mock a Node. However, moking a node is fragile and difficult to maintain, and it is possible that such tests will not benefit.

However, it is quite scary to make any big refactoring until we have no tests. So each big change may produce a lot of bugs, which can not be caught manually in a reasonable time. Especially, since some cases need a synced Node and other things.

I have few ideas about unit testing small things and creating cross-language mocks for GRPC as a testing tool, but it needs more research. If you have ideas — feel free to share.

**Comment by avive** [comment]
I think that all of the proposals are good patterns but in terms of priority tests should be first and break this epic into smaller refactoring tasks that can be handled at times where there are no new major feature in the product pipeline.

**Comment by brusherru** [comment]
It is partially done, and partially stale.
So let's close this one. If we still want to make some of refactoring from this list — let's create a new issue with the narrow scope.
